Trace similarity feature

ck06delhi opened the pull request:
Add functionality for similarity calculation between traces.

ck06delhi committed:
added trace similarity feature

ck06delhi committed:
Improve documentation for trace similarity feature and improve organisation of payloads for testing trace similarity feature.

mohit-at-delhivery commented on assets/docs/functions_documentation.md:
```diff
@@ -205,3 +205,48 @@ clean_trace = clean_output["cleaned_trace"]
+The `calculate_trace_similarity` function enables users to measure the spatial and temporal similarity between two traces. It identifies overlapping pings within specified distance and time thresholds, computes similarity percentages for both traces, and optionally visualizes the traces and their overlap on a map.
+
+#### Parameters
+- **trace_1** *(`list`)*: The first trace, where each ping is represented as `[latitude, longitude, timestamp, ...]`.
```
why the additional "...", the each inner list is only supposed to have latitude, longitude, and timestamp right?

mohit-at-delhivery commented on assets/docs/functions_documentation.md:
```diff
@@ -205,3 +205,48 @@ clean_trace = clean_output["cleaned_trace"]
+
+#### Behavior
+The function returns a dictionary:
+- **`similarity_percentage`** *(`float`)*: The overall similarity percentage between the two traces.
```
Overall word is ambiguous. Please define properly if this is the maximum of similarity (t1,t2) and similarity(t2,t1). Also in the description first explain what do you exactly mean by similarity between (t1,t2) and then explain why these two can be different. 

Also update the key name "similarity_percentage" to "max_similarity_percentage" for more clarity?

mohit-at-delhivery commented on assets/docs/functions_documentation.md:
```diff
@@ -205,3 +205,48 @@ clean_trace = clean_output["cleaned_trace"]
+- **`metadata`** *(`dict`)*:
+  - **`similarity_info_trace_1_to_2`** *(`dict`)*:
+    - **`similarity_percentage`** *(`float`)*: Similarity percentage of `trace_1` with respect to `trace_2`.
+    - **`overlapping_pings_indices`** *(`list`)*: Index pairs of overlapping pings in `trace_1` and `trace_2`.
```
please explain that in an overlapping index (ping_index_from_t1, ping_index_from_t2) pair between t1 and t2, the haversine distance between these two pings is minimum while satisfying the time threshold.

mohit-at-delhivery commented on examples/trace_similarity_example.py:
```diff
@@ -0,0 +1,43 @@
+if __name__ == "__main__":
+
+    # Load existing traces
+    traces_path = constants.BASE_PATH + "/tests/similarity_calculation_payloads/large_trace_pair.json"
```
use load_calculate_trace_similarity_payloads() function here.

mohit-at-delhivery commented on src/tracely/clean_trace.py:
```diff
@@ -1294,3 +1296,174 @@ def plot_raw_vs_stop_comparison_map(self) -> fl.plugins.DualMap:
+        within the given distance and time thresholds. Overlapping pings are paired by their indices, and the similarity percentage is the percentage of pings in `trace_1` that are overlapping.
+
+        Args:
+            trace_1 (np.array): The first trace, represented as a NumPy array, where each row is [latitude, longitude, timestamp].
```
2d numpy array?

mohit-at-delhivery commented on src/tracely/clean_trace.py:
```diff
@@ -1294,3 +1296,174 @@ def plot_raw_vs_stop_comparison_map(self) -> fl.plugins.DualMap:
+        Args:
+            trace_1 (np.array): The first trace, represented as a NumPy array, where each row is [latitude, longitude, timestamp].
+            trace_2 (np.array): The second trace, represented as a NumPy array, where each row is [latitude, longitude, timestamp].
+            distance_threshold (float): The maximum allowable spatial distance (in meters) for two pings to be considered overlapping.
```
remove spatial.

mohit-at-delhivery commented on src/tracely/clean_trace.py:
```diff
@@ -1294,3 +1296,174 @@ def plot_raw_vs_stop_comparison_map(self) -> fl.plugins.DualMap:
+            trace_1 (np.array): The first trace, represented as a NumPy array, where each row is [latitude, longitude, timestamp].
+            trace_2 (np.array): The second trace, represented as a NumPy array, where each row is [latitude, longitude, timestamp].
+            distance_threshold (float): The maximum allowable spatial distance (in meters) for two pings to be considered overlapping.
+            time_threshold (int): The maximum allowable time difference (in the same units as the timestamps, eg. unix epoch timestamp) for two pings to be considered overlapping.
```
please specify milliseconds clearly. no need of mentioning unix etc here.

mohit-at-delhivery commented on src/tracely/clean_trace.py:
```diff
@@ -1294,3 +1296,174 @@ def plot_raw_vs_stop_comparison_map(self) -> fl.plugins.DualMap:
+            time_threshold (int): The maximum allowable time difference (in the same units as the timestamps, eg. unix epoch timestamp) for two pings to be considered overlapping.
+
+        Returns:
+            dict: 
```
why extra empty lines?

mohit-at-delhivery commented on src/tracely/clean_trace.py:
```diff
@@ -1294,3 +1296,174 @@ def plot_raw_vs_stop_comparison_map(self) -> fl.plugins.DualMap:
+                    similarity_percentage (float): The percentage of pings in `trace_1` that have overlapping pings in `trace_2`.
+                    overlapping_pings_indices (list): A list of pairs of indices, where each pair represents the indices of overlapping pings in `trace_1` and `trace_2`.
+
+        Notes:
```
not standard doc string. explain within summary itself.

mohit-at-delhivery commented on src/tracely/clean_trace.py:
```diff
@@ -1294,3 +1296,174 @@ def plot_raw_vs_stop_comparison_map(self) -> fl.plugins.DualMap:
+        overlapping_pings_pairs = []
+
+        # Iterate through pings in trace_1
+        for i, (lat1, lon1, ts1) in enumerate(trace_1):
```
We have used "lng" for longitude everywhere. please maintain same notations. Also rename variables like ts1 to timestamp_1 etc.

mohit-at-delhivery commented on src/tracely/clean_trace.py:
```diff
@@ -1294,3 +1296,174 @@ def plot_raw_vs_stop_comparison_map(self) -> fl.plugins.DualMap:
+
+            # Calculate the difference array
+            candidate_pings_np = np.array(candidate_pings)
+            difference_array = candidate_pings_np - np.array([lat1, lon1])
```
not the right way of calculating the nearest point. calculate haversine distance all candidate pings and query ping then pick the min distance one.

mohit-at-delhivery commented on src/tracely/clean_trace.py:
```diff
@@ -1294,3 +1296,174 @@ def plot_raw_vs_stop_comparison_map(self) -> fl.plugins.DualMap:
+        It computes the similarity percentage for both traces and optionally generates a map visualization of the traces and their overlapping pings.
+
+        Args:
+            trace_1 (list): The first trace as list of list, where each sublist represents [latitude, longitude, timestamp, ...].
```
replace "sublist" with "inner list", and remove extra ... in the list since there are no other expected elements in an inner list.

mohit-at-delhivery commented on src/tracely/clean_trace.py:
```diff
@@ -1294,3 +1296,174 @@ def plot_raw_vs_stop_comparison_map(self) -> fl.plugins.DualMap:
+            trace_1 (list): The first trace as list of list, where each sublist represents [latitude, longitude, timestamp, ...].
+                            Here, latitude and longitude are in degrees, and timestamp is unix epoch timestamp in milliseconds.
+            trace_2 (list): The second trace as list of list, where each sublist represents [latitude, longitude, timestamp, ...].
+                            Here, latitude and longitude are in degrees, and timestamp is unix epoch timestamp in milliseconds.
```
"decimal degrees".

mohit-at-delhivery commented on src/tracely/clean_trace.py:
```diff
@@ -1294,3 +1296,174 @@ def plot_raw_vs_stop_comparison_map(self) -> fl.plugins.DualMap:
+                            Here, latitude and longitude are in degrees, and timestamp is unix epoch timestamp in milliseconds.
+            trace_2 (list): The second trace as list of list, where each sublist represents [latitude, longitude, timestamp, ...].
+                            Here, latitude and longitude are in degrees, and timestamp is unix epoch timestamp in milliseconds.
+            distance_threshold (float): The maximum distance (in meters) within which two pings are considered overlapping.
```
keep one definition fixed for one argument so that it maintains consistency when used across multiple functions.

mohit-at-delhivery commented on src/tracely/clean_trace.py:
```diff
@@ -1294,3 +1296,174 @@ def plot_raw_vs_stop_comparison_map(self) -> fl.plugins.DualMap:
+                            Here, latitude and longitude are in degrees, and timestamp is unix epoch timestamp in milliseconds.
+            distance_threshold (float): The maximum distance (in meters) within which two pings are considered overlapping.
+            time_threshold (int): The maximum allowable time difference in milliseconds for two pings to be considered overlapping.
+            plot_map (bool, optional): A flag indicating whether to generate a folium map visualization of the traces and their overlapping pings. Defaults to `False`.
```
we are not showing overlapping index pairs now due to very large map size right?

mohit-at-delhivery commented on src/tracely/clean_trace.py:
```diff
@@ -1294,3 +1296,174 @@ def plot_raw_vs_stop_comparison_map(self) -> fl.plugins.DualMap:
+                            overlapping_pings_indices (list): Pairs of indices representing overlapping pings in `trace_2` and `trace_1`.
+                    plot (folium.plugins.DualMap or None): A folium map visualization of the traces, if plot_map is True. Otherwise, None.
+
+        Notes:
```
mention in summary.

mohit-at-delhivery commented on src/tracely/clean_trace.py:
```diff
@@ -1294,3 +1296,174 @@ def plot_raw_vs_stop_comparison_map(self) -> fl.plugins.DualMap:
+
+        # Clean traces by removing null values and sorting by timestamp
+        columns = ["latitude", "longitude", "timestamp"]
+        trace_1_df = pd.DataFrame(trace_1, columns=columns)
```
do we really need dataframes just to create a list of non null pings?

mohit-at-delhivery commented on src/tracely/clean_trace.py:
```diff
@@ -1294,3 +1296,174 @@ def plot_raw_vs_stop_comparison_map(self) -> fl.plugins.DualMap:
+        }
+
+        # Plot map according to plot_map argument
+        if plot_map:
+            similarity_percentage_trace_1_to_2 = round(similarity_info_trace_1_to_2["similarity_percentage"], 2)
+            similarity_percentage_trace_2_to_1 = round(similarity_info_trace_2_to_1["similarity_percentage"], 2)
+
+            left_text  = f"<div><b>Trace Similarity (T1, T2)</b>: {similarity_percentage_trace_1_to_2} %</div>"
+            right_text = f"<div><b>Trace Similarity (T2, T1)</b>: {similarity_percentage_trace_2_to_1} %</div>"
```
why isn't this all within the plotting function itself?

mohit-at-delhivery commented on src/tracely/clean_trace.py:
```diff
@@ -1294,3 +1296,174 @@ def plot_raw_vs_stop_comparison_map(self) -> fl.plugins.DualMap:
+        trace_2_np = np.array(clean_trace_2_list)
+        trace_1_representative_location = (np.mean(trace_1_np[:, 0]), np.mean(trace_1_np[:, 1]))
+        trace_2_representative_location = (np.mean(trace_2_np[:, 0]), np.mean(trace_2_np[:, 1]))
+        representative_loc_for_both_traces = ((trace_1_representative_location[0] + trace_2_representative_location[0]) / 2,
```
these can also go within the plotting function itself.

mohit-at-delhivery commented on src/tracely/clean_trace.py:
```diff
@@ -1294,3 +1296,174 @@ def plot_raw_vs_stop_comparison_map(self) -> fl.plugins.DualMap:
+        similarity_result = {"similarity_percentage": similarity_percentage, "metadata": metadata, "plot": plot}
+
+        # Validate output
+        len_clean_trace_1_list = len(clean_trace_1_list)
```
pass these lists itself. calculate lengths within.

mohit-at-delhivery commented on src/tracely/exceptions/error_messages.py:
```diff
@@ -43,7 +43,8 @@ class ValidationErrorMessage:
     INVALID_MIN_SIZE_FOR_STOP_EVENTS_DETECTION = "min_size cannot be less than 2"
     INVALID_MAX_DIST_RATIO_FOR_IMPUTATION = "max_dist_ratio can not be less than 0, but got {}"
-
+    INVALID_TRACE_FOR_OVERLAP_ESTIMATION = "Input traces must contain latitude, longitude and timestamp for all pings."
```
but you have allowed None for latitude, longitude right? isn't this a very generic error message?

mohit-at-delhivery commented on src/tracely/clean_trace.py:
```diff
@@ -1294,3 +1296,174 @@ def plot_raw_vs_stop_comparison_map(self) -> fl.plugins.DualMap:
+            "overlapping_pings_indices": overlapping_pings_pairs
+        }
+
+    @staticmethod
+    def calculate_trace_similarity(trace_1, trace_2, distance_threshold, time_threshold, plot_map=False):
```
This function does not belong to the clean trace class. Let's create a package called "trace_functions" or something and within it we can have "trace_similarity.py" module. This leaves room for adding multiple functionalities here itself.

mohit-at-delhivery commented on src/tracely/utils/data_validation_utils.py:
```diff
@@ -794,4 +794,70 @@ def validate_convert_csv_to_trace_payload_parameters(csv_file_path,
+            ValidationException (DATA_FORMAT_EXCEPTION_CODE: 4002): If `trace_2` is not of data type list.
+            ValidationException (DATA_FORMAT_EXCEPTION_CODE: 4002): If a ping in `trace_1` is not of data type list.
+            ValidationException (DATA_FORMAT_EXCEPTION_CODE: 4002): If a ping in `trace_2` is not of data type list.
+
```
missing "plot_map" parameter exception

mohit-at-delhivery commented on src/tracely/utils/data_validation_utils.py:
```diff
@@ -794,4 +794,70 @@ def validate_convert_csv_to_trace_payload_parameters(csv_file_path,
+
+            # Validate presence of latitude, longitude and timestamp
+            if len(ping) < 3:
+                raise ValidationException(ValidationErrorMessage.INVALID_TRACE_FOR_OVERLAP_ESTIMATION,
```
INVALID_LENGTH_OF_PING_FOR_OVERLAP_ESTIMATION?

mohit-at-delhivery commented on src/tracely/utils/data_validation_utils.py:
```diff
@@ -794,4 +794,70 @@ def validate_convert_csv_to_trace_payload_parameters(csv_file_path,
+            
+        for ping in trace_2:
+            # Check if ping is an empty list
+            DataValidationUtils.check_empty_list(ping, "ping")
+
+            # Validate presence of lat, lng, timestamp
+            if len(ping) < 3:
+                raise ValidationException(ValidationErrorMessage.INVALID_TRACE_FOR_OVERLAP_ESTIMATION,
+                                        ValidationErrorCode.VALUE_EXCEPTION_CODE)
+
+            # Validate latitude, longitude, timestamp
+            DataValidationUtils.check_latitude(ping[0], "latitude")
+            DataValidationUtils.check_longitude(ping[1], "longitude")
+            DataValidationUtils.check_timestamp(ping[2], "timestamp")
```
can have a small function for checking one ping in overlap_estimation since these lines are repeated.

mohit-at-delhivery commented on tests/test_script.py:
```diff
@@ -1742,6 +1749,262 @@ def test_interpolate_trace_invalid_combination_for_min_and_max_dist_from_prev_pi
+    with pytest.raises(ValidationException, match=expected_error_msg):
+        CleanTrace.calculate_trace_similarity(trace_1, 
+                                              trace_2, 
+                                              distance_threshold = 0, # Invalid value: negative
```
wrong comment

mohit-at-delhivery commented on tests/test_script.py:
```diff
@@ -1742,6 +1749,262 @@ def test_interpolate_trace_invalid_combination_for_min_and_max_dist_from_prev_pi
+    with pytest.raises(ValidationException, match=expected_error_msg):
+        CleanTrace.calculate_trace_similarity(trace_1, 
+                                              trace_2, 
+                                              distance_threshold = 1, # Invalid type: string
```
wrong comment

mohit-at-delhivery commented on tests/test_script.py:
```diff
@@ -1742,6 +1749,262 @@ def test_interpolate_trace_invalid_combination_for_min_and_max_dist_from_prev_pi
+        CleanTrace.calculate_trace_similarity(trace_1, 
+                                              trace_2, 
+                                              distance_threshold = 1, # Invalid type: string
+                                              time_threshold = 1, plot_map = "invalid")  
```
keep plot_map in a new line

mohit-at-delhivery commented on tests/test_script.py:
```diff
@@ -1742,6 +1749,262 @@ def test_interpolate_trace_invalid_combination_for_min_and_max_dist_from_prev_pi
+        CleanTrace.calculate_trace_similarity(trace_1, 
+                                              trace_2, 
+                                              distance_threshold = 1, 
+                                              time_threshold = 1)  # Invalid type: string
```
wrong comment

mohit-at-delhivery commented on assets/docs/Tracely I_O Structure - output of calculate_trace_similarity method.pdf:
write in clean trace output format.

mohit-at-delhivery reviewed (commented):
Kindly incorporate the suggested changes. Thank you.

mohit-at-delhivery committed:
INCORPORATE REVIEW SUGGESTIONS:
- Improve documentation
- Use dedicated function to load dummy traces
- Extend test cases
- Trace similarity calculation does not belong to CleanTrace class, move it to dedicated module as independent function.
- Calculate dependencies within function rather than calculating them in parent function and then passing as parameter. (eg: right_text, left_text in plotting functionality.)
